src/hotspot/cpu/x86/x86.ad

```diff
@@ -7527,26 +7539,41 @@ instruct rearrangeB_evex(vec dst, vec src, vec shuffle) %{
 
 instruct loadShuffleS(vec dst, vec src, vec vtmp, rRegP scratch) %{
   predicate(vector_element_basic_type(n) == T_SHORT &&
-            vector_length(n) <= 8 && !VM_Version::supports_avx512bw()); // NB! aligned with rearrangeS
+            vector_length(n) <= 16 && !VM_Version::supports_avx512bw()); // NB! aligned with rearrangeS
   match(Set dst (VectorLoadShuffle src));
   effect(TEMP dst, TEMP vtmp, TEMP scratch);
   format %{ "vector_load_shuffle $dst, $src\t! using $vtmp and $scratch as TEMP" %}
   ins_encode %{
     // Create a byte shuffle mask from short shuffle mask
     // only byte shuffle instruction available on these platforms
+    int vlen_in_bytes = vector_length_in_bytes(this);
+    if (UseAVX == 0) {
```

Comment:
This code will not be executed with vector length 16 because match_rule_supported_vector() bailout with (size_in_bits == 256 && UseAVX < 2).

Reply:
Yes you are right, but this code will execute for vector length 16 when UseAVX ==2.
 It will also execure for vector length 16 when UseAVX == 3 &&
!VM_Version::supports_avx512bw.

Reply:
Next assert checks <= 16 when code is guarded by (UseAVX == 0). It is not (UseAVX ==2).
Also } else { case is for UseAVX > 0 which includes AVX=1 but vpaddb() (avx3) is used there.
Seems UseAVX checks wrong here.

Reply:
The assert checks for  vlen_in_bytes <= 16 (128 bits) and so is a correct check for UseAVX=0.
vpaddb is supported on  AVX1/AVX2 as well.
vpaddb is supported on AVX1 for up to 128 bit and
on AVX2 for upto 256 bit and 
on AVX3 (512) for upto 512 bit vectors.
I have tested this for UseAVX=0, UseAVX=1, UseAVX=2, UseAVX=3 platform.

The check is for UseAVX as with any flavor of AVX, we can use less number of instructions to do this operation.
This is because AVX allows destination to be separate from both the sources. 

Please let me know if I am missing something.

Reply:
My bad - I missed that size is in bytes in assert. The assert is correct, as you said.
And `} else {` part works for AVX1 because of match_rule_supported_vector() bailout 256-bit case.
May be add assert(UseAVX > 1 || vlen_in_bytes <= 16, ).

I only have one question left - about check >= 256 in  match_rule_supported_vector()

Reply:
Added the following assert on else path:
+      assert(UseAVX > 1 || vlen_in_bytes <= 16, "required");

Reply:
Good.